src/zkevm_specs/evm_circuit/execution/error_gas_uint_overflow.py

```diff
@@ -0,0 +1,117 @@
+from ...util import (
+    FQ,
+    TxDataNonZeroGasEIP2028,
+    MAX_U64,
+    N_BYTES_U64,
+    TxGas,
+    TxGasContractCreation,
+    TxDataZeroGas,
+)
+from ..instruction import Instruction, Transition
+from ..table import CallContextFieldTag
+from ..opcode import Opcode
+
+
+def error_gas_uint_overflow(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    (
+        is_call,
+        is_callcode,
+        is_delegatecall,
+        is_staticcall,
+        is_create_flag,
+        is_create2_flag,
+    ) = instruction.multiple_select(
+        opcode,
+        (
+            Opcode.CALL,
+            Opcode.CALLCODE,
+            Opcode.DELEGATECALL,
+            Opcode.STATICCALL,
+            Opcode.CREATE,
+            Opcode.CREATE2,
+        ),
+    )
+    is_create = is_create_flag + is_create2_flag
+
+    # init overflow flag
+    is_call_gas_cost_overflow = (
+        is_eip2028_overflow
+    ) = is_non_zero_gas_overflow = is_eip3860_overflow = FQ(0)
+
+    # call gas_cost overflow flag.
+    # seems never overflow because of checking range inside of CallGadget
+    tx_id = instruction.call_context_lookup(CallContextFieldTag.TxId)
+    call = CallGadget(instruction, FQ(0), is_call, is_callcode, is_delegatecall)
+    is_warm_access = instruction.read_account_to_access_list(tx_id, call.callee_address)
+    gas_cost = call.gas_cost(instruction, is_warm_access)
+    is_call_gas_cost_overflow = is_call * instruction.is_u64_overflow(gas_cost)
+
+    # intrinsic gas flag.
+    calldata_offset = instruction.call_context_lookup(CallContextFieldTag.CallDataOffset)
+    calldata_length = instruction.call_context_lookup(CallContextFieldTag.CallDataLength)
+    data = [
+        instruction.tx_calldata_lookup(tx_id, calldata_offset + FQ(idx))
+        for idx in range(calldata_length.expr().n)
+    ]
+    dataLen = len(data)
+
+    def non_zero_gas_constraints():
+        # eip 2028
+        nz = len([byte for byte in data if byte != 0])
+        gas = TxGasContractCreation if is_create == FQ(1) else TxGas
+        non_zero_gas = TxDataNonZeroGasEIP2028
+        is_eip2028_overflow, _ = instruction.compare(
+            FQ(((MAX_U64 - gas) // non_zero_gas)), FQ(nz), N_BYTES_U64
+        )
+        gas += nz * non_zero_gas
+
+        # tx data zero gas overflow
+        z = dataLen - nz
+        zero_gas = TxDataZeroGas
+        is_non_zero_gas_overflow, _ = instruction.compare(
+            FQ(((MAX_U64 - gas) // zero_gas)), FQ(z), N_BYTES_U64
+        )
+        gas += z * zero_gas
+
+        # TODO: eip 3860
+        # if is_create:
+        #     lenWords = dataLen // 32
+        #     is_eip3860_overflow, _ = instruction.compare(
+        #         FQ((MAX_U64 - gas) // InitCodeWordGas), FQ(lenWords), N_BYTES_U64
+        #     )
+
+    instruction.condition(FQ(dataLen > 0), non_zero_gas_constraints)
+
+    # verify gas uint overflow.
+    is_overflow = (
```

Comment:
As Adria mentioned, I didn't see how `CALLDATACOPY`, `CODECOPY`, `EXTCODECOPY` and `RETURNDATACOPY` were handled. Bcs this error could happen in many different places, using different functions to handle different error cases is a good idea, just like `non_zero_gas_constraints()`.